packages/gatsby/src/schema/index.js

```diff
@@ -19,10 +19,54 @@ const getAllFieldExtensions = () => {
   }
 }
 
-const build = async ({ parentSpan }) => {
+// Schema building requires metadata for type inference.
+// Technically it means looping through all type nodes, analyzing node structure
+// and then using this aggregated node structure in related GraphQL type.
+// Actual logic for inference located in inferenceMetadata reducer and ./infer
+// Here we just orchestrate the process via redux actions
+const buildInferenceMetadata = ({ types }) =>
+  new Promise(resolve => {
+    if (!types || !types.length) {
+      resolve()
+      return
+    }
+    const typeNames = [...types]
+    // TODO: use async iterators when we switch to node>=10
+    //  or better investigate if we can offload metadata building to worker/Jobs API
+    //  and then feed the result into redux?
+    const processNextType = () => {
+      const typeName = typeNames.pop()
+      store.dispatch({
+        type: `BUILD_TYPE_METADATA`,
+        payload: {
+          typeName,
+          nodes: nodeStore.getNodesByType(typeName),
+        },
+      })
+      if (typeNames.length > 0) {
+        // Give event-loop a break
+        setTimeout(processNextType, 0)
```

Comment:
Is there need to do that?

Reply:
This whole process can take seconds (for big sites). And since it is synchronous it will block the event loop and even make CLI "freeze" for that time. So I applied it as a [recommended best practice](https://nodejs.org/en/docs/guides/dont-block-the-event-loop/#complex-calculations-without-blocking-the-event-loop). Do you afraid of some race conditions / conflicts with redux actions in between?

Reply:
That was my assumption that's the case, because this is called during bootstrap only so we are (more or less) in control what's being executed, so I didn't see "functional" reason to do that (and letting cli do the "spinner tick" periodically makes sense)